pylot/drivers/carla_collision_sensor_operator.py

```diff
@@ -0,0 +1,118 @@
+        self._logger.debug("@[{}]: Received Vehicle ID: {}".format(
+            vehicle_id_msg.timestamp, vehicle_id))
+
+        # Connect to the world.
+        _, world = get_world(self._flags.carla_host, self._flags.carla_port,
+                             self._flags.carla_timeout)
+        if world is None:
+            raise ValueError("There was an issue connecting to the simulator.")
+
+        self._vehicle = get_vehicle_handle(world, vehicle_id)
+
+        # Install the collision sensor.
+        collision_blueprint = world.get_blueprint_library().find(
+            'sensor.other.collision')
+        self._logger.debug("Spawning a collision sensor.")
+        self._collision_sensor = world.spawn_actor(collision_blueprint,
+                                                   carla.Transform(),
+                                                   attach_to=self._vehicle)
+
+        # Register the callback on the collision sensor.
+        self._collision_sensor.listen(self.process_collision)
+
+        # Register an on_tick function to flow watermarks.
+        world.on_tick(self.process_timestamp)
+
+    def process_collision(self, collision_event):
+        """ Invoked when a collision event is received from the simulation.
+
+        Args:
+            collision_event (:py:class:`carla.CollisionEvent`): A collision
+                event that contains the impulse, location and the object with
+                which the ego-vehicle collided.
+        """
+        game_time = int(collision_event.timestamp * 1000)
+        self._logger.debug(
+            "@[{}]: Received a collision event from the simulator.".format(
+                game_time))
+
+        # Create a CollisionMessage.
+        timestamp = erdos.Timestamp(coordinates=[game_time])
+        msg = CollisionMessage(
+            collision_event.other_actor,
+            Vector3D.from_carla_vector(collision_event.normal_impulse),
+            timestamp)
+
+        # Send the CollisionMessage.
+        self._collision_stream.send(msg)
+
+    def process_timestamp(self, msg):
+        """ Invoked upon each tick of the simulator. This callback is used to
+        flow watermarks to the downstream operators.
+
+        Since multiple collision events can be published by Carla for a single
+        timestamp, we need to wait till the next tick of the simulator to be
+        sure that all the collision events for the previous timestamp have
+        been sent to downstream operators.
+
+        Args:
+            msg (:py:class:`carla.WorldSettings`): A snapshot of the world at
+            the given tick. We use this to retrieve the timestamp of the
+            simulator.
+        """
+        timestamp = erdos.Timestamp(
+            coordinates=[int(msg.elapsed_seconds * 1000)])
+        if self._timestamp_to_close:
+            last_timestamp = self._timestamp_to_close
+            self._collision_stream.send(erdos.WatermarkMessage(last_timestamp))
```

Comment:
Just to double check. I assume it's not an issue that we delay the watermark msg by one tick because no operator on the critical path depends on this collision operator. Is that right?

Reply:
Yeah, it's just a logger, so we'll wait an extra tick before pushing stuff to the logs, but that's fine. We still get all collisions on time, so if some future operator wants to actually work on this data, they can work on messages instead of watermarks.